### ISSUE TITLE
indirect_structural_match lint misfiring on enum variant unused in const item

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#62614]
(Spawned off of https://github.com/rust-lang/rust/issues/62411#issuecomment-510604193)

Consider this code ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=449daf68183e9566fc106b070fdaeb10)):

```rust
struct Sum(u32, u32);

impl PartialEq for Sum {
    fn eq(&self, other: &Self) -> bool { self.0 + self.1 == other.0 + other.1 }
}

impl Eq for Sum { }

#[derive(PartialEq, Eq)]
enum Eek {
    TheConst,
    UnusedByTheConst(Sum)
}

const THE_CONST: Eek = Eek::TheConst;

pub fn main() {
    match Eek::UnusedByTheConst(Sum(1,2)) {
        THE_CONST => { println!("Hello"); }
        _ => { println!("Gbye"); }
    }
}
```

In stable and beta, it compiles without warning.

In nightly, it compiles with the following diagnostic warning (#62411):

```
warning: to use a constant of type `Sum` in a pattern, `Sum` must be annotated with `#[derive(PartialEq, Eq)]`
  --> src/main.rs:19:9
   |
19 |         THE_CONST => { println!("Hello"); }
   |         ^^^^^^^^^
   |
   = note: #[warn(indirect_structural_match)] on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #62411 <https://github.com/rust-lang/rust/issues/62411>
```

This is probably not what we want for this lint: It is entirely reasonable for the user to assume that the structural_match check is based solely on the ADT's that one encounters when traversing *the const expression itself*, and one should ignore any unused enum variants that are only tangentially related to the const.

**Comment by pnkfelix** [comment]
My current thinking is that maybe we should revert PR #62339.

 * The "right way" to do this, I think, isn't going to use a type-visitor at all, so the code there is going to be thrown away anyway.

 * I'm not going to have time to fix it properly today, and after today I'm going to be gone for 8 weeks, which means if nothing is done, the current behavior will propagate to beta.

 * So the only question is whether the current lint, with its false positives, provides significantly more value to our users than the negative effects that its false positives have on our users.

**Comment by pnkfelix** [comment]
New thought (since PR #62339 *did* include some useful fixes to e.g. ICEs): don't revert, "just" change the lint to default to `Allow` for now. We can move its default to `Warn` after I fix #62614

**Comment by pnkfelix** [comment]
when/if PR #62623 lands, the stable-to-nightly regression aspect of this issue can be removed.
